sotodlib/coords/det_match.py

```diff
@@ -0,0 +1,670 @@
+from typing import List, Optional, Tuple, Iterator
+from copy import deepcopy
+import h5py
+
+import numpy as np
+import matplotlib.pyplot as plt
+from scipy.optimize import linear_sum_assignment
+
+from sotodlib.coords import optics
+from sotodlib.core import metadata
+from sotodlib.io.metadata import write_dataset
+
+warnings.filterwarnings(action='ignore',
+                        message='overflow encountered in exp*', category=RuntimeWarning)
+
+
+def map_band_chans(b1, c1, b2, c2, chans_per_band=512):
+    """
+    Returns an index mapping of length nchans1 from one set of bands and
+    channels to another. Note that unmapped indices are returned as -1, so this
+    must be handled before (or after) indexing or else you'll get weird
+    results.
+    Args:
+        b1 (np.ndarray):
+            Array of length nchans1 containing the smurf band of each channel
+        c1 (np.ndarray):
+            Array of length nchans1 containing the smurf channel of each channel
+        b2 (np.ndarray):
+            Array of length nchans2 containing the smurf band of each channel
+        c2 (np.ndarray):
+            Array of length nchans2 containing the smurf channel of each channel
+        chans_per_band (int):
+            Lets just hope this never changes.
+    """
+    acs1 = b1 * chans_per_band + c1
+    acs2 = b2 * chans_per_band + c2
+
+    mapping = np.full_like(acs1, -1, dtype=int)
+    for i, ac in enumerate(acs1):
+        x = np.where(acs2 == ac)[0]
+        if len(x > 0):
+            mapping[i] = x[0]
+    return mapping
+
+
+@dataclass
+class Resonator:
+    """
+    Data structure to hold any resonator information.
+    """
+    idx: int
+    res_freq: float
+    is_north: int
+
+    res_qi: float = np.nan
+    smurf_res_idx: int = -1
+    smurf_band: int = -1
+    smurf_channel: int = -1
+    bg: int = -1
+
+    xi: float = np.nan
+    eta: float = np.nan
+    gamma: float = np.nan
+
+    det_x: float = np.nan
+    det_y: float = np.nan
+    det_rhomb: str = ''
+    det_pol: str = ''
+    det_freq: int = 0
+    det_bandpass: str = ''
+    is_optical: int = 1
+
+    mux_band: int = -1
+    mux_channel: int = -1
+    mux_layout_pos: int = -1
+
+    matched: int = 0
+    match_idx: int = -1
+ 
+
+def apply_design_properties(src_res, dst_res, in_place=False):
+    """
+    Combines two resonators into one, taking smurf-properties such as res-idx,
+    smurf-band, and smurf-channel one, and design properties such as det
+    position and polarization from the other.
+
+    Args:
+        src_res (Resonator):
+            The resonator to take smurf properties from
+        dst_res (Resonator):
+            The resonator to take design properties from
+        in_place (bool):
+            If True, the src_res will be modified. Otherwise, a new Resonator
+            will be created and returned.
+    """
+    if in_place:
+        r = src_res
+    else:
+        r = deepcopy(src_res)
+    r.det_x = src_res.det_x
```

Comment:
I think you want dst_res on RHS of all the lines from here down, right?  Either that or the construction of r is wrong...